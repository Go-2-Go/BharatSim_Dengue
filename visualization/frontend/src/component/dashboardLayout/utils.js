import React from 'react';
import renderChart from '../charts/renderChart';

export function renderElement(element) {
  return (
    <div key={element.i} data-grid={element} data-testid={element.i}>
      <div style={{ height: '100%', width: '100%', boxSizing: 'border-box', padding: '10px' }}>
        {renderChart(element.chartType, { config: element.config })}
      </div>
    </div>
  );
}

export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 2) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 2,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
