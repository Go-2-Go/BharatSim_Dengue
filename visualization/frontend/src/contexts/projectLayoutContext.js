import React from 'react';

const projectLayoutContext = React.createContext({
  projectMetadata: {
    name:'',
  },
  selectedDashboardMetadata: {
    name: '',
  },
});

const ProjectLayoutProvider = projectLayoutContext.Provider;

export { projectLayoutContext, ProjectLayoutProvider };
