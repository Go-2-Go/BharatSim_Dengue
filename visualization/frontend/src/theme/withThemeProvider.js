import React from 'react';
import { MuiThemeProvider, CssBaseline } from '@material-ui/core';

import theme from './theme';

function withThemeProvider(WrappedComponent) {
  return function WithTheme() {
    return (
      <MuiThemeProvider theme={theme}>
        <CssBaseline />
        <WrappedComponent />
      </MuiThemeProvider>
    );
  };
}

export default withThemeProvider;
