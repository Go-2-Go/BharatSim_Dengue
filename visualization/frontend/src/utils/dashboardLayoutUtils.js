/* eslint-disable react/prop-types */

import React from 'react';
import renderChart from '../component/charts/renderChart';

export function renderElement({ layout, chartType, config }) {
  return (
    <div key={layout.i} data-grid={layout} data-testid={layout.i}>
      <div style={{ height: '100%', width: '100%', boxSizing: 'border-box', padding: '10px' }}>
        {renderChart(chartType, { config })}
      </div>
    </div>
  );
}

export function getNewWidgetLayout(numberOfWidgetAdded, cols, count) {
  return createLayout({
    id: `widget-${count}`,
    xPosition: (numberOfWidgetAdded * 2) % cols,
    yPosition: Infinity, // puts it at the bottom
    width: 2,
    height: 2,
  });
}

function createLayout({ id, xPosition, yPosition, width, height }) {
  return {
    i: id,
    x: xPosition,
    y: yPosition,
    w: width,
    h: height,
  };
}
