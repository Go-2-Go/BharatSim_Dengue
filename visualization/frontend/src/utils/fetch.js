import axios from 'axios';
import { httpMethods } from '../constants/fetch';

let shouldShowError;

function commonErrorHandler(err) {
  shouldShowError(true);
}

function initApiConfig({ setError }) {
  shouldShowError = setError;
}

function fetchData({ url, method = httpMethods.GET, headers, data, query, customErrorHandler }) {
  return axios({ url, method, headers, data, params: query })
    .then((res) => res.data)
    .catch((err) => {
      if (customErrorHandler) {
        customErrorHandler(err);
        return;
      }
      commonErrorHandler(err);
    });
}

function uploadData({ url, headers, data, query, customErrorHandler }) {
  return fetchData({
    url,
    method: httpMethods.POST,
    headers,
    data,
    params: query,
    customErrorHandler,
  });
}

export { fetchData, uploadData, initApiConfig };
