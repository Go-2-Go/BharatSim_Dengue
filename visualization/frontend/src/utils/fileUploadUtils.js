const fileUploadedStatus = {
  LOADING: 'loading',
  SUCCESS: 'success',
  ERROR: 'error',
};

function getStatusAndMessageFor(fileUploadStatus, fileName) {
  return {
    [fileUploadedStatus.ERROR]: {
      status: fileUploadedStatus.ERROR,
      message: `Error occurred while unloading ${fileName}`,
    },
    [fileUploadedStatus.SUCCESS]: {
      status: fileUploadedStatus.SUCCESS,
      message: `${fileName} successfully uploaded`,
    },
    [fileUploadedStatus.LOADING]: {
      status: fileUploadedStatus.LOADING,
      message: `uploading ${fileName}`,
    },
  }[fileUploadStatus];
}

export { fileUploadedStatus, getStatusAndMessageFor };
