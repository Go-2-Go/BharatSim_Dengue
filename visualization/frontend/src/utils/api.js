import { contentTypes, httpMethods } from '../constants/fetch';
import { url as serviceURL } from './url';
import { fetchData, uploadData } from './fetch';

function headerBuilder({ contentType }) {
  return {
    'content-type': contentType,
  };
}

function formDataBuilder(data) {
  const formData = new FormData();
  data.forEach(({ name, value }) => formData.append(name, value));
  return formData;
}

const api = {
  saveDashboard: async ({ widgets, layout, dashboardId, name, count }) => {
    return uploadData({
      url: serviceURL.DASHBOARD_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
      data: JSON.stringify({
        dashboardData: { widgets, layout, dashboardId, name, count },
      }),
    });
  },

  getAllDashBoard: async () => fetchData({ url: serviceURL.DASHBOARD_URL }),

  uploadFileAndSchema: async ({ file, schema }) =>
    uploadData({
      url: serviceURL.DATA_SOURCES,
      data: formDataBuilder([
        { name: 'datafile', value: file },
        { name: 'schema', value: JSON.stringify(schema) },
      ]),
      headers: headerBuilder({ contentType: contentTypes.FILE }),
    }),

  getCsvHeaders: async (dataSourceId) => fetchData({ url: serviceURL.getHeaderUrl(dataSourceId) }),

  getDatasources: async () => fetchData({ url: serviceURL.DATA_SOURCES }),

  getData: async (datasource, query) =>
    fetchData({ url: serviceURL.getDataUrl(datasource), query }),

  saveProject: async ({ id, ...data }) => {
    const requestObject = {
      url: serviceURL.PROJECT_URL,
      headers: headerBuilder({ contentType: contentTypes.JSON }),
    };
    if (id) {
      return uploadData({
        ...requestObject,
        data: JSON.stringify({ projectData: { ...data, id } }),
        method: httpMethods.PUT,
      });
    }
    return uploadData({
      ...requestObject,
      data: JSON.stringify({ projectData: { ...data } }),
      method: httpMethods.POST,
    });
  },

  getProjects: async () => {
    return fetchData({
      url: serviceURL.PROJECT_URL,
    });
  },
  getProject: async (id) => {
    return fetchData({
      url: serviceURL.getProjectUrl(id),
    });
  },
};

export { api };
