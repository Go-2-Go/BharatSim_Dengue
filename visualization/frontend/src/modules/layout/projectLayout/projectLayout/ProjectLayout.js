import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

import { Box } from '@material-ui/core';
import LoaderOrError from '../../../../component/loaderOrError/LoaderOrError';
import SideDashboardNavbar from '../sideDashboardNavbar/SideDashboardNavbar';

import useProjectLayoutStyle from './projectLayoutCSS';
import useFetch from '../../../../hook/useFetch';

import { api } from '../../../../utils/api';
import { ProjectLayoutProvider } from '../../../../contexts/projectLayoutContext';
import { ChildrenPropTypes } from '../../../../commanPropTypes';

function ProjectLayout({ children }) {
  const classes = useProjectLayoutStyle();
  const { id } = useParams();

  const [projectMetadata, setProjectMetadata] = useState({
    id: undefined,
    name: 'untitled project',
  });
  const [dashboards, setDashboards] = useState([]);
  const [selectedDashboard] = useState(0);

  const { data: fetchedProjectMetadata, loadingState: projectLoadingState } = useFetch(
    api.getProject,
    [id],
    !!id,
  );

  const { data: fetchedDashboards, loadingState: dashboardLoadingState } = useFetch(
    api.getAllDashBoardByProjectId,
    [id],
    !!id,
  );

  useEffect(() => {
    if (fetchedDashboards) setDashboards(fetchedDashboards.dashboards);
  }, [fetchedDashboards]);

  useEffect(() => {
    if (fetchedProjectMetadata) {
      const { _id, name } = fetchedProjectMetadata.project;
      setProjectMetadata({ id: _id, name });
    }
  }, [fetchedProjectMetadata]);

  return (
    <Box className={classes.layoutContainer}>
      <Box className={classes.sideBarLayout}>
        <SideDashboardNavbar
          navItems={dashboards.map((dashboard) => dashboard.name)}
          value={selectedDashboard}
        />
      </Box>
      <Box display="flex" flex={1} flexDirection="column">
        <LoaderOrError loadingState={projectLoadingState}>
          <LoaderOrError loadingState={dashboardLoadingState}>
            <ProjectLayoutProvider
              value={{
                projectMetadata,
                selectedDashboardMetadata: dashboards[selectedDashboard] || {},
              }}
            >
              {children}
            </ProjectLayoutProvider>
          </LoaderOrError>
        </LoaderOrError>
      </Box>
    </Box>
  );
}

ProjectLayout.propTypes = {
  children: ChildrenPropTypes.isRequired,
};

export default ProjectLayout;
