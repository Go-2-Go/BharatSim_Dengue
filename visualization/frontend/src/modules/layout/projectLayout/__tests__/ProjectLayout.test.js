import React, { useContext } from 'react';
import { render } from '@testing-library/react';
import { fireEvent, waitFor } from '@testing-library/dom';
import * as router from 'react-router-dom';
import { api } from '../../../../utils/api';
import withThemeProvider from '../../../../theme/withThemeProvider';
import ProjectLayout from '../projectLayout/ProjectLayout';
import { projectLayoutContext } from '../../../../contexts/projectLayoutContext';

const mockHistoryPush = jest.fn();
const mockHistoryReplace = jest.fn();

function DummyComponent() {
  const { projectMetadata, selectedDashboardMetadata } = useContext(projectLayoutContext);
  return <div>{JSON.stringify({ projectMetadata, selectedDashboardMetadata }, null, 2)}</div>;
}

jest.mock('../../../../utils/api', () => ({
  api: {
    getProject: jest.fn(),
    saveProject: jest.fn(),
  },
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useParams: jest.fn(),
  useHistory: () => ({
    push: mockHistoryPush,
    replace: mockHistoryReplace,
  }),
}));

describe('Project', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const Component = withThemeProvider(() => (
    <ProjectLayout>
      <DummyComponent />
    </ProjectLayout>
  ));

  it('should match snapshot while creating new project', async () => {
    router.useParams.mockReturnValue({ id: undefined });

    const { container, findByText } = render(<Component />);

    await findByText('untitled project');

    expect(container).toMatchSnapshot();
  });

  it('should match snapshot for old projects', async () => {
    router.useParams.mockReturnValue({ id: 1 });
    api.getProject.mockResolvedValue({ project: { _id: 1, name: 'project1' } });

    const { container, findByText } = render(<Component />);

    await findByText('project1');

    expect(container).toMatchSnapshot();
  });

  it('should show error components for failure while fetching project', async () => {
    router.useParams.mockReturnValue({ id: 1 });
    api.getProject.mockRejectedValue();

    const { findByText, queryByText } = render(<Component />);

    await findByText('Failed to load, Refresh the page');

    expect(queryByText('Failed to load, Refresh the page')).toBeInTheDocument();
  });

  it('should update the url on successful saving of project', async () => {
    router.useParams.mockReturnValue({ id: undefined });
    api.saveProject.mockResolvedValue({ projectId: 1 });
    api.getProject.mockResolvedValue({ project: { _id: 1, name: 'project1' } });

    const { getByText, findByText } = render(<Component />);

    await findByText('untitled project');

    fireEvent.click(getByText('Save'));

    router.useParams.mockReturnValue({ id: 1 });

    await waitFor(() =>
      expect(mockHistoryReplace).toHaveBeenCalledWith({ pathname: '/projects/1' }),
    );
  });

  it('should save project on click of save  button for old project', async () => {
    api.saveProject.mockResolvedValue({ projectId: 1 });
    api.getProject.mockResolvedValue({ project: { _id: 1, name: 'project1' } });
    router.useParams.mockReturnValue({ id: 1 });
    const { getByText, findByText } = render(<Component />);

    await findByText('Save');
    fireEvent.click(getByText('Save'));

    expect(api.saveProject).toHaveBeenCalledWith({ id: 1, name: 'project1' });
  });
});
