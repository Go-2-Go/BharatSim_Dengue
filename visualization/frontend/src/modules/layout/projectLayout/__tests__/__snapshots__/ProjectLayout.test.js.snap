// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project should render child without any api call if project id is present 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-23 makeStyles-layoutContainer-21"
  >
    <div
      class="MuiBox-root MuiBox-root-24 makeStyles-sideBarLayout-22"
    >
      <div
        class="MuiBox-root MuiBox-root-31"
      >
        <div
          class="MuiTabs-root MuiTabs-vertical"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              class="MuiTabs-flexContainer MuiTabs-flexContainerVertical"
              role="tablist"
            >
              <button
                aria-selected="true"
                class="MuiButtonBase-root MuiTab-root makeStyles-root-25 MuiTab-textColorInherit makeStyles-textColorInherit-29 Mui-selected makeStyles-selected-28 MuiTab-labelIcon makeStyles-labelIcon-27 MuiTab-fullWidth"
                data-testid="tab-d_name"
                role="tab"
                tabindex="0"
                type="button"
              >
                <span
                  class="MuiTab-wrapper makeStyles-wrapper-26"
                >
                  <div
                    class="MuiBox-root MuiBox-root-40"
                  >
                    <img
                      alt="dashboard-logo"
                      src="dashboard-icon.svg"
                    />
                  </div>
                  d_name
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <span
              class="PrivateTabIndicator-root-35 PrivateTabIndicator-colorSecondary-37 MuiTabs-indicator makeStyles-indicator-30 PrivateTabIndicator-vertical-38"
              style="top: 0px; height: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-32"
    >
      <div>
        <div>
          ProjectLayout Child
        </div>
        {
  "projectMetadata": {
    "id": 1,
    "name": "project1"
  },
  "selectedDashboardMetadata": {
    "name": "d_name",
    "_id": "d_id"
  }
}
      </div>
    </div>
  </div>
</div>
`;

exports[`Project should render child without any api call if project id is undefined 1`] = `
<div>
  <div
    class="MuiBox-root MuiBox-root-4 makeStyles-layoutContainer-2"
  >
    <div
      class="MuiBox-root MuiBox-root-5 makeStyles-sideBarLayout-3"
    >
      <div
        class="MuiBox-root MuiBox-root-12"
      >
        <div
          class="MuiTabs-root MuiTabs-vertical"
        >
          <div
            class="MuiTabs-scroller MuiTabs-fixed"
            style="overflow: hidden;"
          >
            <div
              class="MuiTabs-flexContainer MuiTabs-flexContainerVertical"
              role="tablist"
            />
            <span
              class="PrivateTabIndicator-root-16 PrivateTabIndicator-colorSecondary-18 MuiTabs-indicator makeStyles-indicator-11 PrivateTabIndicator-vertical-19"
              style="top: 0px; height: 0px;"
            />
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root MuiBox-root-13"
    >
      <div>
        <div>
          ProjectLayout Child
        </div>
        {
  "projectMetadata": {
    "name": "untitled project"
  },
  "selectedDashboardMetadata": {}
}
      </div>
    </div>
  </div>
</div>
`;
